_posts/2020-11-22-catch2.md

```diff
@@ -0,0 +1,644 @@
+
+## ¡Tarda mucho en compilar!
+
+No me queda la más mínima duda de que habrás ido copiando todos los ejemplos del post, compilandolos y ejecutándolos. Por supuesto. Claro.
+
+Y te habrás dado cuenta de que tardan bastante en compilar. Francamente, es un rollazo.
+
+Pero es el precio a pagar por tener todo en un único fichero de cabecera.
+
+Sin embargo, hay una solución.
+
+Te creas un fichero main.cpp (puedes cambiar el nombre) con este contenido:
+
+```
+#define CATCH_CONFIG_MAIN
+#include <catch2/catch.hpp>
+```
+
+Fíjate que en este fichero dejamos CATCH_CONFIG_MAIN.
+
+Y, por otro lado, te creas un fichero cpp por cada conjunto significativo de tests. Por ejemplo, p008.cpp:
+
+```
+#include <catch2/catch.hpp>
+
+TEST_CASE() {}
+```
+Fíjate que en estos ficheros omitimos CATCH_CONFIG_MAIN.
+
+Ahora los compilas por separado y los enlazas juntos:
+```
+$ g++ -I . -c main.cpp
+$ g++ -I . -c p008.cpp 
+$ g++ -o p008 main.o p008.o
+```
+
+(Por supuesto, usa CMake, pero no te voy a dar ahora un curso de eso, ¿no?)
+
+Compilar main.cpp seguirá llevando bastante tiempo, pero no hace falta que lo recompiles cada dos por tres, porque nunca lo cambiarás.
+
+Estarás cambiando el resto de ficheros de tests. Pero esos tardan muy muy poco en recompilarse gracias a la omisión de CATCH_CONFIG_MAIN.
+
+## Matchers
+
+Con la macro REQUIRE deberías cubrir el 80% de tus necesidades. Por no decir el 99%.
+
+Sin embargo hay veces que las expansiones de las aserciones dejan bastante que desear.
+
+Pongamos que quieres escribir una aserción que compruebe que la cadena de texto devuelta por una función contiene una subcadena dada:
+
+```
+#define CATCH_CONFIG_MAIN
+#include <catch2/catch.hpp>
+#include <string>
+
+std::string f() {
+    return "Tres trostes tigres";
+}
+
+TEST_CASE("Los tigres deberían estar tristes") {
+    REQUIRE( f().find("tristes") != std::string::npos );
+}
+```
+
+Ejecutemos:
+```
+$ ./p009
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+p009 is a Catch v2.13.3 host application.
+Run with -? for options
+
+-------------------------------------------------------------------------------
+Los tigres deberían estar tristes
+-------------------------------------------------------------------------------
+p009.cpp:9
+...............................................................................
+
+p009.cpp:10: FAILED:
+  REQUIRE( f().find("tristes") != std::string::npos )
+with expansion:
+  18446744073709551615 (0xffffffffffffffff)
+  !=
+  18446744073709551615 (0xffffffffffffffff)
+
+===============================================================================
+test cases: 1 | 1 failed
+assertions: 1 | 1 failed
+```
+
+Vale, vale. Supongo que si lo piensas un poco tiene sentido, pero... joer, ¡vaya porquería de mensaje de error!
+
+Lo que necesitamos es una forma de especificarle a Catch el mensaje que queremos que se imprima cuando la aserción falla.
```

Comment:
No, lo que necesitamos es una forma de decirle lo que realmente quiere decir el test.

Hago esta puntualizacion porque en este caso concreto, y lo que intentan los matchers, lo que se intenta es reprensentar el test con las herramientas del framework (De forma que "sabe" dar una representacion natural del error). Esto es muy diferente a dar soporte a tipos de terceros para que Catch los pinte. Por ejemplo:

``` cpp
MyString foo, bar:

REQUIRE(foo, EndWith(bar));
```

esto daria un mensaje horripilante del tipo:

```
...
with expansion:
<MyString object at 0x55445343> ends with <MyString object at 0x123123242>
```

Reply:
Ummm.

Según la documentación oficial "matchers are an alternative way to write assertions, useful for tests where you work with complex types or need to assert more complex properties. Matchers are easily composable and users can write their own and combine them with the Catch2-provided matchers seamlessly."

Desde luego no indican que los Matchers sean la forma de especificar los mensajes de error. Eso lo tengo que redactar de otra forma.

Así que creo que tienes razón con el comentario. Pero yo me estaba queriendo centrar en que los mensajes a veces son horripilantes y los matchers ayudan en eso.

Lo voy a redactar otra vez.

Reply:
Reescrito.